### PR TITLE
v3.18.0-SNAPSHOT exchange rate and account models [release-type snapshot]

### PR DESCRIPTION
**Pull request by TimElectrum**
### New Features
* Added new models:
- `ExchangeRate` to describe the exchange rate between two currencies.
- `Account` to describe more varied accounts in a consistent manner.
This has the following sub-types defined:
- `BankAccount`
- `IbanAccount`
- `IfsAccount`
- `SwiftAccount`
- `MobileWalletAccount`
* The limitations on the `id` field of the `Institution` model have
been removed to make the field suitable for a wider range of
applications. The values in this field need no longer be defined by
Electrum and similar values as recognised at third parties may now be
carried in this field. However, care should be taken during
implementation to ensure that different parties agree on the values
which will be present in these fields.

### CODE REVIEW

File: src/main/java/io/electrum/vas/model/Account.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+   @NotNull
+   public String getAccountId() {
+      return accountId;
+   }
+
+   public void setAccountId(String accountId) {
+      this.accountId = accountId;
+   }
+
+   @ApiModelProperty(required = true, value = "The specific type of account. By default this is simply ACCOUNT which only caries an account number.")
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/Account.java]: Typo: Carries
Fun fact: Apparently dental caries is another term for cavities (tooth decay).

File: src/main/java/io/electrum/vas/model/MobileWalletAccount.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package io.electrum.vas.model;
+
+import java.util.Objects;
+
+import io.electrum.vas.Utils;
+import io.swagger.annotations.ApiModel;
+
+@ApiModel(description = "A model to describe mobiel wallet accounts. Such accounts consist of just an account number. This account number is typically the customer's MSISDN.")
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/MobileWalletAccount.java]: typo on mobile

File: src/main/java/io/electrum/vas/model/IfsAccount.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package io.electrum.vas.model;
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/IfsAccount.java]: I made a typo in the spec, should be IFSC (Indian Financial System Code). However, this is essentially the same as a swift bank transfer, except that IFSC is for India, whereas swift is for whomever implements it.

File: src/main/java/io/electrum/vas/model/ExchangeRate.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+/**
+ * An object describing the exchange rate from one currency to another.
+ **/
+@ApiModel(description = "An object describing the exchange rate from one currency to another.")
+public class ExchangeRate {
+
+   protected float rate;
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/ExchangeRate.java]: Did you try BigDecimal?

File: src/main/java/io/electrum/vas/model/ExchangeRate.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+   @Override
+   public int hashCode() {
+      return Objects.hash(rate, fromCurrency, toCurrency);
+   }
+
+   @Override
+   public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("class ExchangeRate {\n");
+
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/ExchangeRate.java]: random new line

File: src/docs/devguide/hugo/site/content/release-notes.md [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+## 3.18.0
+Released TBC February 2019
+
+### New Features
+* Added new models:
+  - `ExchangeRate` to describe the exchange rate between two currencies.
+  - `Account` to describe more varied accounts in a consistent manner. This has the following sub-types defined:
+    - `BankAccount`
+    - `IbanAccount`
+    - `IfsAccount`
```
Comment (VeeneshJeena) [inline src/docs/devguide/hugo/site/content/release-notes.md]: IFSC

File: src/main/java/io/electrum/vas/model/IfsAccount.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.Objects;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import io.electrum.vas.Utils;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "A model to describe IFS accounts")
```
Comment (VeeneshJeena) [inline src/main/java/io/electrum/vas/model/IfsAccount.java]: IFSC (Indian Financial System Code)

File: src/main/java/io/electrum/vas/model/IfsAccount.java (hunk 3 above)
Comment (VeeneshJeena) [reply src/main/java/io/electrum/vas/model/IfsAccount.java]: I think we did decide that we were going to separate them, so don't pay too much attention to the second sentence.

File: src/main/java/io/electrum/vas/model/ExchangeRate.java (hunk 4 above)
Comment (VeeneshJeena) [reply src/main/java/io/electrum/vas/model/ExchangeRate.java]: Should we mention the accuracy (amount of decimal places) in the API? Or is this an implementation thing?

Comment (TimElectrum) [reply src/main/java/io/electrum/vas/model/ExchangeRate.java]: no, i think that's an implementation thing. Different languages all do different things at tiny precision levels.

**Pull request merged**